# Lab 3: Contextual Bandit-Based News Article Recommendation

**`Course`:** Reinforcement Learning Fundamentals  
**`Student Name`:**  
**`Roll Number`:**  
**`GitHub Branch`:** firstname_U20230xxx  

# Imports and Setup

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score

from rlcmab_sampler import sampler


# Load Datasets

In [2]:
# Load datasets
news_df = pd.read_csv("data/news_articles.csv")
train_users = pd.read_csv("data/train_users.csv")
test_users = pd.read_csv("data/test_users.csv")

news_df.head()

link  \
0  https://www.huffpost.com/entry/covid-boosters-...   
1  https://www.huffpost.com/entry/american-airlin...   
2  https://www.huffpost.com/entry/funniest-tweets...   
3  https://www.huffpost.com/entry/funniest-parent...   
4  https://www.huffpost.com/entry/amy-cooper-lose...   

                                            headline   category  \
0  Over 4 Million Americans Roll Up Sleeves For O...  U.S. NEWS   
1  American Airlines Flyer Charged, Banned For Li...  U.S. NEWS   
2  23 Of The Funniest Tweets About Cats And Dogs ...     COMEDY   
3  The Funniest Tweets From Parents This Week (Se...  PARENTING   
4  Woman Who Called Cops On Black Bird-Watcher Lo...  U.S. NEWS   

                                   short_description               authors  \
0  Health experts said it is too early to predict...  Carla K. Johnson, AP   
1  He was subdued by passengers and crew when he ...        Mary Papenfuss   
2  "Until you have a dog you don't understand wha...         Elyse Wanshel   
3  "Accidentally put grown-up toothpaste on my to...      Caroline Bologna   
4  Amy Cooper accused investment firm Franklin Te...        Nina Golgowski   

         date  
0  2022-09-23  
1  2022-09-23  
2  2022-09-23  
3  2022-09-23  
4  2022-09-22

In [3]:
train_users.head()

user_id   age  income  clicks  purchase_amount  session_duration  \
0   U7392   NaN   23053      10           500.00             17.34   
1   U2702  56.0   20239      11           913.33             22.22   
2   U2461   NaN   13907       9          1252.62             41.57   
3   U7475   NaN   26615      12           500.00             30.17   
4   U6040  32.0   27958      13           500.00             65.27   

   content_variety  engagement_score  num_transactions  avg_monthly_spend  \
0          0.36661          37.29781                 3             187.44   
1          0.61370          59.36342                 5             145.15   
2          0.80368          76.78706                 7             282.03   
3          0.26499          30.19441                10             195.35   
4          0.36385          37.12153                 5             439.68   

   ...  screen_brightness  battery_percentage  cart_abandonment_count  \
0  ...                4.0                 2.0                       8   
1  ...                4.5                63.0                       5   
2  ...                1.3                22.0                       2   
3  ...                4.2                77.0                       9   
4  ...                4.6                30.0                       9   

   browser_version  background_app_count  session_inactivity_duration  \
0          3.17.97                    10                        22.75   
1          1.57.10                     8                         1.75   
2          2.16.94                    12                        29.33   
3          9.90.20                     4                        21.61   
4          1.99.38                     7                         7.58   

   network_jitter  region_code  subscriber   label  
0             8.0         Z999       False  user_3  
1             4.0         U428        True  user_2  
2            18.0         Z999        True  user_3  
3            22.0         X123       False  user_3  
4            52.0         S043       False  user_1  

[5 rows x 33 columns]

## Data Preprocessing

In this section:
- Handle missing values
- Encode categorical features
- Prepare data for user classification

In [4]:
for name, df in {
    "news": news_df,
    "train_users": train_users,
    "test_users": test_users
}.items():
    print(f"\n{name}")
    print(f"Total entries: {len(df)}")
    na_counts = df.isna().sum()
    na_counts = na_counts[na_counts > 0]
    if not na_counts.empty:
        print(na_counts)

print(f'\n{(news_df["category"] == "").sum()=}')
print(f'{(train_users["label"] == "").sum()=}')


news
Total entries: 209527
headline                 6
short_description    19712
authors              37418
dtype: int64

train_users
Total entries: 2000
age    698
dtype: int64

test_users
Total entries: 2000
age    679
dtype: int64

(news_df["category"] == "").sum()=np.int64(0)
(train_users["label"] == "").sum()=np.int64(0)


**Null Value Handling**
1. All the news articles do have a non null or empty category.
2. Filling in the missing age as the median value of the entire dataset.

In [5]:
train_users['age'] = train_users['age'].fillna(train_users['age'].median())
test_users['age'] = test_users['age'].fillna(test_users['age'].median())

In [6]:
# Find non-numeric columns in train_users and test_users
non_numeric_train_users = train_users.select_dtypes(include='object').columns.tolist()
non_numeric_test_users = test_users.select_dtypes(include='object').columns.tolist()

print("Non-numeric columns in train_users:", non_numeric_train_users)
print("Non-numeric columns in test_users:", non_numeric_test_users)

Non-numeric columns in train_users: ['user_id', 'browser_version', 'region_code', 'label']
Non-numeric columns in test_users: ['user_id', 'browser_version', 'region_code']


/var/folders/lx/8dqwd2x96b3gmdmvm7rg5tr00000gn/T/ipykernel_64569/4123833300.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  non_numeric_train_users = train_users.select_dtypes(include='object').columns.tolist()
/var/folders/lx/8dqwd2x96b3gmdmvm7rg5tr00000gn/T/ipykernel_64569/4123833300.py:3: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silenc

In [7]:
print(f"{train_users['browser_version'].nunique()=}")
print(f"{train_users['region_code'].nunique()=}")
print(f"{train_users['label'].nunique()=}")

train_users['browser_version'].nunique()=1973
train_users['region_code'].nunique()=1284
train_users['label'].nunique()=3


**Drop Columns and Encode**
- Droppping `user_id` as it is not needed for classification.
- There are 2000 training points with 1973 unique browser version, therefore dropping `browser_version` as well.
- Dropping `region_code` as well as 1284 unique variation.
- `subscriber` already encoding as True / False or 0 / 1

In [8]:
columns_to_drop = ['user_id', 'browser_version', 'region_code']

train_users = train_users.drop(columns=columns_to_drop, errors='ignore')
test_users = test_users.drop(columns=columns_to_drop, errors='ignore')

## User Classification

Train a classifier to predict the user category (`User1`, `User2`, `User3`),
which serves as the **context** for the contextual bandit.

> The classifier code can be found in `classifiers.py`

In [10]:
import classifiers as myclsf

# Drop user_id col 
train_users = train_users.drop(columns=["user_id"], errors='ignore')

X_train, y_train, X_test, y_test, lr = myclsf.preprocess(train_users)

results, models = myclsf.train_classifiers(X_train, y_train, X_test, y_test)

results

/opt/homebrew/Caskroom/miniforge/base/envs/mlpr/lib/python3.13/site-packages/sklearn/linear_model/_logistic.py:406: ConvergenceWarning: lbfgs failed to converge after 10000 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=10000).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


{'Logistic Regression': 0.73,
 'Random Forest': 0.79,
 'SVM (RBF)': 0.46,
 'KNN': 0.555,
 'Neural Network': 0.3625}

After trying multiple methods, **Random Forest** has the best accuracy, and will be used as the classifier for subsequest tasks.

---

# Contextual Bandit

## Reward Sampler Initialization

The sampler is initialized using the student's roll number `i`.
Rewards are obtained using `sampler.sample(j)`.


## Arm Mapping

| Arm Index (j) | News Category | User Context |
|--------------|---------------|--------------|
| 0–3          | Entertainment, Education, Tech, Crime | User1 |
| 4–7          | Entertainment, Education, Tech, Crime | User2 |
| 8–11         | Entertainment, Education, Tech, Crime | User3 |

## Epsilon-Greedy Strategy

This section implements the epsilon-greedy contextual bandit algorithm.


## Upper Confidence Bound (UCB)

This section implements the UCB strategy for contextual bandits.

## SoftMax Strategy

This section implements the SoftMax strategy with temperature $ \tau = 1$.


## Reinforcement Learning Simulation

We simulate the bandit algorithms for $T = 10,000$ steps and record rewards.

P.S.: Change $T$ value as and if required.


## Results and Analysis

This section presents:
- Average Reward vs Time
- Hyperparameter comparisons
- Observations and discussion


## Final Observations

- Comparison of Epsilon-Greedy, UCB, and SoftMax
- Effect of hyperparameters
- Strengths and limitations of each approach
